libbeat/outputs/elasticsearch/client_test.go

```diff
@@ -713,6 +713,93 @@ func BenchmarkCollectPublishFailAll(b *testing.B) {
 	}
 }
 
+func BenchmarkPublish(b *testing.B) {
+	requestCount := 0
+
+	// start a mock HTTP server
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(b, "testing value", r.Header.Get("X-Test"))
+		// from the documentation: https://golang.org/pkg/net/http/
+		// For incoming requests, the Host header is promoted to the
+		// Request.Host field and removed from the Header map.
+		assert.Equal(b, "myhost.local", r.Host)
+
+		var response string
+		if r.URL.Path == "/" {
+			response = `{ "version": { "number": "7.6.0" } }`
+		} else {
+			response = `{"items":[{"index":{}},{"index":{}},{"index":{}}]}`
+
+		}
+		fmt.Fprintln(w, response)
+		requestCount++
+	}))
+	defer ts.Close()
+
+	event := beat.Event{
+		Timestamp: time.Date(2017, time.November, 7, 12, 0, 0, 0, time.UTC),
+		Meta: mapstr.M{
+			"a.b": "c",
+			"metaLevel0Map": mapstr.M{
+				"metaLevel1Map": mapstr.M{
+					"metaLevel2Value": "metavalue3",
+					"new1":            "newmetavalue1",
+				},
+			},
+			"metaLevel0Value":  "metareplaced1",
+			"metaLevel0Value2": "untouched",
+			"new2":             "newmetavalue2",
+		},
+		Fields: mapstr.M{
+			"a.b": "c",
+			"fieldsLevel0Map": mapstr.M{
+				"fieldsLevel1Map": mapstr.M{
+					"fieldsLevel2Value": "fieldsvalue3",
+					"new3":              "newfieldsvalue1",
+				},
+				"newmap": mapstr.M{
+					"new4": "newfieldsvalue2",
+				},
+			},
+			"fieldsLevel0Value":  "fieldsreplaced1",
+			"fieldsLevel0Value2": "untouched",
+		},
+	}
+
+	levels := []int{1, 4, 7, 9}
+
+	// Indexing to _bulk api
+	// It uses gzip encoder internally for encoding data
+	for _, l := range levels {
+		b.Run(fmt.Sprintf("level %d", l), func(b *testing.B) {
+			client, err := NewClient(
+				clientSettings{
+					connection: eslegclient.ConnectionSettings{
+						URL: ts.URL,
+						Headers: map[string]string{
+							"host":   "myhost.local",
+							"X-Test": "testing value",
+						},
+						CompressionLevel: l,
+					},
+				},
+
+				nil,
+			)
+			assert.NoError(b, err)
+			batch := encodeBatch(client, outest.NewBatch(event, event, event))
```

Comment:
I don't think testing just a batch of 3 relatively small events gives us enough data.

We need to compare small batches, medium batches and large batches, let's say:
* 5 events for a small
* 50 events for a medium
* 500 for a large

It's possible that one compression algorithm is better than the other at a particular size and compression level.